### ISSUE TITLE
Broken with Gatsby 2.18.0

### ISSUE THREAD

**Issue by wKovacs64** [graysonhicks/gatsby-plugin-remote-images#28]
Something seems to have changed in Gatsby [`2.17.17 -> 2.18.0`](https://github.com/gatsbyjs/gatsby/compare/gatsby@2.17.17...gatsby@2.18.0) that impacts this plugin. I'm no longer seeing the node created. Confirmed missing via GraphiQL, and the build error is as follows:

```
 ERROR #85907  GRAPHQL

There was an error in your GraphQL query:

- Unknown field 'siteImageFile' on type 'Site'.
```

[CodeSandbox Reproduction](https://codesandbox.io/s/gatsby-plugin-remote-images-and-gatsby-2180-07imc)

**Comment by avwie** [comment]
I was pulling my hair out, because I have the exact same thing. My `localImage` node is not showing anymore. This broke my complete build.

**Comment by Zebusch** [comment]
Seeing the same thing. Downgrading to 2.17, it works just fine.

**Comment by wKovacs64** [comment]
This seems related to https://github.com/gatsbyjs/gatsby/pull/19092. I did some debugging of this plugin and `node` has the new `myImageFile___NODE` field on it at the end of `onCreateNode`, so Gatsby must be stripping or ignoring it or something?

cc @vladar

**Comment by vladar** [comment]
Sorry to see this change broke your plugin. But looks like you are doing direct node mutation in `onCreateNode` which is not supported by Gatsby. Happens here:

https://github.com/graysonhicks/gatsby-plugin-remote-images/blob/649abaa2e88f3c5c5f0e27c2f9cb7bea503b7dc5/src/gatsby-node.js#L72-L76

There is a separate action for situations when you need adding a field to a node: [createNodeField](https://www.gatsbyjs.org/docs/actions/#createNodeField). A quote from docs:

```
Also since nodes are immutable, you can’t mutate the node directly.
So to extend another node, use this.
```

But a better way to declare a foreign key relationship on a node would be using [schema customization APIs](https://www.gatsbyjs.org/docs/schema-customization/) (Gatsby will probably shift towards this way of customizing types in future and deprecate old inference tricks eventually). I think `createResolvers` could be the way to go. Something along the lines:

```js
exports.createResolvers = ({ createResolvers }) => {
  const resolvers = {
    [nodeType]: {
      [fieldName]: {
        type: ["File"],
        resolve(source, args, context, info) {
          return context.nodeModel.getNodeById(source.parentNodeId)
        },
      },
    },
  }
  createResolvers(resolvers)
}
```

Hope it helps!

**Comment by jbolda** [comment]
@vladar I don't know that the direct node manipulation is necessarily the issue the root issue. I am also getting errors related to the `gatsby-source-filesystem` function `createRemoteFileNode` in the `gatsby-source-airtable` plugin. I am getting `- Unknown field 'Attachments' on type 'AirtableData'.` where the plugin does not create this type. The [plugin does use the `___NODE`](https://github.com/jbolda/gatsby-source-airtable/blob/43d77af3ae81c4535f54398608f1c2daf69c4977/gatsby-node.js#L307) (without direct node manipulation because it is a source plugin)/.

Related PR that breaks on ^2.18.0: https://github.com/jbolda/gatsby-source-airtable/pull/117
Job line error: https://travis-ci.com/jbolda/gatsby-source-airtable/jobs/259371321#L510

**Comment by vladar** [comment]
@jbolda Your link points to `localFiles___NODE` on `AirtableField` but the error is about `Attachments` on type `AirtableData`.

I did a quick sanity check - added:

```js
console.log(`create node `, row.id ,` data: `, Object.keys(node.data))
```
before [this `createNode` call](https://github.com/jbolda/gatsby-source-airtable/blob/master/gatsby-node.js#L167) (which will create `Airtable` and `AirtableData` types).

And turns out that the field `Attachments` is not there (see [the log in gist](https://gist.github.com/vladar/6ce51cef0d7661dfb7dde76c9e51bfbf)). It is added somewhere else **after** the `createNode` call is processed by Gatsby. Maybe via mutation [here](https://github.com/jbolda/gatsby-source-airtable/blob/73e7b077af115743b5624c6b237fd65684f25f08/gatsby-node.js#L200-L202) or [here](https://github.com/jbolda/gatsby-source-airtable/blob/master/gatsby-node.js#L245-L247).

There might be some race condition that didn't manifest itself before the schema rebuilding PR but shows up now.

So to summarize this - it is likely a mutation issue. Does it help?

**Comment by kpennell** [comment]
whew! Thank god for github issues !

**Comment by jbolda** [comment]
@vladar It looks like it was actually the `Attachments` returning as a promise which prior gatsby would let resolve, but now `await`ing that promise fixed the bug. The mutations you pointed out are just local to those functions and prior to the node being created, so it doesn't seem that was the issue.

ref: https://github.com/jbolda/gatsby-source-airtable/pull/119/commits/7d84a2d45a69fd79bbb2d0a706274879b7f140f9

Looking at this plugin, it does make the mutations directly on the node, as pointed out, which is likely the issue here. All promises are`await`ed in this plugin. So it seems the two issues are unrelated.

**Comment by lawrenceong001** [comment]
hi totally new to gatsby -- just out of curiosity how do I downgrade back to 2.7? I tried 

`npm unnistall -g gatsby-cli`
`npm i -g gatsby-cli@2.7.59`

on my site, i still end up with 2.8.14

tks

**Comment by ooloth** [comment]
You’re changing the gatsby installation that lives globally on your computer (that’s what the -g flag does).

Instead, you want to change the gatsby installation specific to your project. 

So, navigate your console to your project folder and run `npm i gatsby@2.17.17` (no need to uninstall first).

**Comment by lawrenceong001** [comment]
thanks @ooloth! that solved my image issue. 

tks
Lawrence

**Comment by wKovacs64** [comment]
You can also update to `gatsby@2.18.5`, as a "fix" has landed there (actually it just re-hides the problem, as it was hidden prior to 2.18.0). This plugin (and others) still need updating, but the problem is masked in the latest version of Gatsby.

**Comment by lawrenceong001** [comment]
thanks for the suggestion @wKovacs64. I will look into this after I run through my work with the project. 

tks
Lawrence

**Comment by btiwaree** [comment]
just spent a good 4 hours trying to figure out what was wrong with my local plugin that used `gatsby-plugin-remote-images` 😢, thanks a lot @wKovacs64 for this.

**Comment by graysonhicks** [comment]
Sorry all, for some reason I was not getting my Github notifications on this!  😅 If there is a PR that can improve this plugin to make the move long term to use the `createResolvers` method, I am happy to accept.

@vladar @wKovacs64

**Comment by wKovacs64** [comment]
I took a stab at this today, but I'm running into a situation (race condition?) where sometimes the resulting node from `await createRemoteFileNode(...)` has a non-empty `children` array (this indicates `childImageSharp` is present and everything works) and other times it's an empty array. 🤔

**Comment by graysonhicks** [comment]
> You can also update to `gatsby@2.18.5`, as a "fix" has landed there (actually it just re-hides the problem, as it was hidden prior to 2.18.0). This plugin (and others) still needs updating, but the problem is masked in the latest version of Gatsby.

Long term fix in [PR here](#30 ) by @wKovacs64 .
